les/fetcher.go

```diff
@@ -618,28 +723,78 @@ func (f *lightFetcher) checkAnnouncedHeaders(fp *fetcherPeerInfo, headers []*typ
 // checkSyncedHeaders updates peer's block tree after synchronisation by marking
 // downloaded headers as known. If none of the announced headers are found after
 // syncing, the peer is dropped.
-func (f *lightFetcher) checkSyncedHeaders(p *peer) {
+func (f *lightFetcher) checkSyncedHeaders(p *peer) (bool, []*types.Header, *big.Int) {
 	fp := f.peers[p]
 	if fp == nil {
 		p.Log().Debug("Unknown peer to check sync headers")
-		return
+		return false, nil, nil
 	}
+
+	var h *types.Header
+	if f.pm.isULCEnabled() {
+		var unapprovedHashes []common.Hash
+		// Overwrite last announced for ULC mode
+		h, unapprovedHashes = f.lastValidTrieNode(p)
+		//rollback untrusted blocks
+		f.chain.Rollback(unapprovedHashes)
+	}
+
 	n := fp.lastAnnounced
 	var td *big.Int
+	trustedHeaderExisted := false
+
+	//find last trusted block
 	for n != nil {
-		if td = f.chain.GetTd(n.hash, n.number); td != nil {
+		//we found last trusted header
+		if n.hash == h.Hash() {
+			trustedHeaderExisted = true
+		}
+		if td = f.chain.GetTd(n.hash, n.number); td != nil && trustedHeaderExisted {
+			break
+		}
+
+		//break if we found last trusted hash before sync
+		if n.hash == f.getLastValidHeader().Hash() {
 			break
 		}
 		n = n.parent
 	}
-	// now n is the latest downloaded header after syncing
-	if n == nil {
+
+	// Now n is the latest downloaded/approved header after syncing
+	if n == nil && !p.isTrusted {
 		p.Log().Debug("Synchronisation failed")
 		go f.pm.removePeer(p.id)
-	} else {
-		header := f.chain.GetHeader(n.hash, n.number)
-		f.newHeaders([]*types.Header{header}, []*big.Int{td})
+		return false, nil, nil
+	}
+	header := f.chain.GetHeader(n.hash, n.number)
+	return true, []*types.Header{header}, td
+}
+
+// lastValidTrieNode return last approved treeNode and a list of unapproved hashes
+func (f *lightFetcher) lastValidTrieNode(p *peer) (*types.Header, []common.Hash) {
```

Comment:
It's not a trie node, just a tree (fetcher tree) node. Trie is the name of the Merkle tree. Also, it is not "valid" but "trusted", just like in case of setLastValidHeader.
Also, the implementation is not right, iterating back on the canonical chain you might never reach the last head because a reorg could have rendered it non-canonical since then. core/rawdb.FindCommonAncestor can find the last common ancestor which can definitely be trusted. I would not call it directly on lastTrustedHeader and currentHeader because currentHeader might be far above lastTrustedHeader after syncing and iterating back can be slow. You can efficiently find the common ancestor like this:
```
canonical := f.chain.CurrentHeader()
if canonical.number.Uint64() > f.lastTrustedHeader.Number.Uint64() {
  canonical = f.chain.GetHeaderByNumber(f.lastTrustedHeader.Number.Uint64())
}
ancestorHash := rawdb.FindCommonAncestor(f.pm.chainDb, canonical, f.lastTrustedHeader).Hash()
```
Then you can roll back until you find ancestorHash or a trusted hash.